applications/crossbar/src/crossbar_doc.erl

```diff
@@ -156,10 +158,65 @@ load([_|_]=IDs, Context, Opts, _RespStatus) ->
         {'ok', JObjs} ->
             Docs = extract_included_docs(JObjs),
-            cb_context:store(handle_couch_mgr_success(Docs, Context)
-                             ,'db_doc'
-                             ,Docs
-                            )
+            case check_document_type(IDs, Context, Docs, Opts) of
+                'true' ->
+                    cb_context:store(handle_couch_mgr_success(Docs, Context), 'db_doc', Docs);
+                'fasle' ->
+                    ErrorCause = wh_json:from_list([{<<"cause">>, IDs}]),
+                    cb_context:add_system_error('bad_identifier', ErrorCause, Context)
+            end
+    end.
+
+-spec get_open_function(wh_proplist()) -> function().
+get_open_function(Options) ->
+    case props:get_is_true('use_cache', Options, 'true') of
+        'true' ->  fun couch_mgr:open_cache_doc/3;
+        'false' -> fun couch_mgr:open_doc/3
+    end.
+
+-spec check_document_type(api_binary(), cb_context:context(), wh_json:object(), wh_proplist()) ->
+                                  boolean().
+check_document_type(DocId, Context, JObj, Options) when is_binary(DocId) ->
+    JObjType = wh_doc:type(JObj),
+    ExpectedType = props:get_value(?OPTION_EXPECTED_TYPE, Options),
+    [{Noun, _}| _] = cb_context:req_nouns(Context),
+    ReqType = normalize_requested_resource_name(Noun),
+    document_type_match(JObjType, ExpectedType, ReqType);
+check_document_type([], _Context, _JObjs, _Options) ->
+    'true';
+check_document_type([DocId|DocIds], Context, [JObj|JObjs], Options) ->
+    case check_document_type(DocId, Context, JObj, Options) of
+        'true' -> check_document_type(DocIds, Context, JObjs, Options);
+        'false' -> 'false'
+    end.
+
+-spec document_type_match(api_binary(), api_binary(), ne_binary()) -> boolean().
+document_type_match('undefined', _ExpectedType, _ReqType) ->
+    lager:debug("document doesn't have type, requested type is ~p", [_ReqType]),
+    'true';
+document_type_match(_JObjType, 'any', _) -> 'true';
+document_type_match(ExpectedType, ExpectedType, _) -> 'true';
+document_type_match(ReqType, _, ReqType) ->
+    lager:debug("expected type is not specified, checking against requested type ~p", [ReqType]),
+    'true';
+document_type_match(_JObjType, _ExpectedType, _ReqType) ->
+    lager:warning("the document type ~s does not match the expected type ~s nor the requested type ~s"
+              ,[_JObjType, _ExpectedType, _ReqType]),
+    'false'.
+
+-spec normalize_requested_resource_name(ne_binary()) -> ne_binary().
+normalize_requested_resource_name(Name) ->
+    case props:get_value(Name, ?SPECIAL_EXPECTED_TYPE) of
+        'undefined' -> depluralize_resource_name(Name);
+        Special -> Special
+    end.
+
+-spec depluralize_resource_name(ne_binary()) -> ne_binary().
+depluralize_resource_name(Name) ->
+    Size = byte_size(Name) - 1,
+    case Name of
+        <<Bin:Size/binary, L:8>> when L == $s -> Bin;
```

Comment:
I believe you can match `<<Bin:Size, "s">>` now that BEAM knows the size.

(yes it's pretty dumb, binaries "know" their size, so why doesn't BEAM allow rhs matching…)